src/net/socket.rs

```diff
@@ -0,0 +1,222 @@
+                }
+                _ => unreachable!(),
+            }
+        }
+        Ok(())
+    }
+
+    /// Serializes and sends a `Packet` on the socket. On success, returns the number of bytes written.
+    fn send_to(&mut self, packet: Packet) -> NetworkResult<usize> {
+        let connection = self
+            .connections
+            .get_or_insert_connection(&packet.addr(), self.config.clone());
+        let mut packet_data =
+            connection.process_outgoing(packet.payload(), packet.delivery_method())?;
+        let mut bytes_sent = 0;
+
+        if let Some(link_conditioner) = &self.link_conditioner {
+            if link_conditioner.should_send() {
+                for payload in packet_data.parts() {
+                    bytes_sent += self.send_packet(&packet.addr(), &payload)?;
+                }
+            }
+        } else {
+            for payload in connection.gather_dropped_packets() {
+                bytes_sent += self.send_packet(&packet.addr(), &payload)?;
+            }
+
+            for payload in packet_data.parts() {
+                bytes_sent += self.send_packet(&packet.addr(), &payload)?;
+            }
+        }
+
+        Ok(bytes_sent)
+    }
+
+    /// Receives a single message from the socket. On success, returns the packet containing origin and data.
+    fn recv_from(&mut self) -> NetworkResult<Option<Packet>> {
+        let (recv_len, address) = self.socket.recv_from(&mut self.recv_buffer)?;
+        if recv_len == 0 {
+            return Err(NetworkErrorKind::ReceivedDataToShort)?;
+        }
+
+        let received_payload = &self.recv_buffer[..recv_len];
+        let connection = self
+            .connections
+            .get_or_insert_connection(&address, self.config.clone());
+        connection.process_incoming(received_payload)
+    }
+
+    /// Send a single packet over the udp socket.
+    fn send_packet(&self, addr: &SocketAddr, payload: &[u8]) -> NetworkResult<usize> {
+        let mut bytes_sent = 0;
+
+        bytes_sent += self
+            .socket
+            .send_to(payload, addr)
+            .map_err(|io| NetworkError::from(NetworkErrorKind::IOError(io)))?;
+
+        Ok(bytes_sent)
+    }
+
+    fn new(
+        socket: mio::net::UdpSocket,
+        config: NetworkConfig,
+    ) -> (Self, mpsc::Sender<Packet>, mpsc::Receiver<SocketEvent>) {
+        let (event_sender, event_receiver) = mpsc::channel();
+        let (packet_sender, packet_receiver) = mpsc::channel();
+        let buffer_size = config.receive_buffer_max_size;
+        (
+            Self {
+                socket,
+                config: Arc::new(config),
+                connections: ActiveConnections::new(),
+                recv_buffer: vec![0; buffer_size],
+                link_conditioner: None,
+                event_sender,
+                packet_receiver,
+            },
+            packet_sender,
+            event_receiver,
+        )
+    }
+}
+
+impl Evented for LaminarSocket {
```

Comment:
Shouldn't this be documented: `register`,  `reregister` and `deregister`  because they are callable from the Api

Reply:
they are documented here https://docs.rs/mio/0.6.16/mio/event/trait.Evented.html so no need